src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseMessage.cs

```diff
@@ -322,19 +322,12 @@ private static byte[] DecodeSignature(CborReader reader)
         private static List<CoseSignature> DecodeCoseSignaturesArray(CborReader reader, byte[] bodyProtected)
         {
             int? signaturesLength = reader.ReadStartArray();
+            List<CoseSignature> signatures = new List<CoseSignature>(signaturesLength ?? 1);
 
-            if (signaturesLength.GetValueOrDefault() < 1)
-            {
-                throw new CryptographicException(SR.Format(SR.DecodeErrorWhileDecoding, SR.MultiSignMessageMustCarryAtLeastOneSignature));
-            }
-
-            List<CoseSignature> signatures = new List<CoseSignature>(signaturesLength!.Value);
-
-            for (int i = 0; i < signaturesLength; i++)
+            while (reader.PeekState() == CborReaderState.StartArray)
             {
                 int? length = reader.ReadStartArray();
-
-                if (length != CoseMultiSignMessage.CoseSignatureArrayLength)
+                if (length != null && length != CoseMultiSignMessage.CoseSignatureArrayLength)
                 {
                     throw new CryptographicException(SR.Format(SR.DecodeErrorWhileDecoding, SR.DecodeCoseSignatureMustBeArrayOfThree));
```

Comment:
A definite array that is the wrong length will get a nice exception message.

An indefinite array that is the wrong length will presumably get a CBOR exception of some sort.

The best way I'm seeing to unify on the informative message, off the top of my head, would be to keep peeking for EndArray before (or in) the field reader helper methods.  Or to just precount the values for indefinite arrays.

(Right now precounting looks to be hard with CborReader, since it doesn't have `PeekEncodedValue`, but we could add one)

Reply:
I don't think is worth adding extra logic for validation of indefinite-length arrays, I don't envision that they will be commonly used for COSE objects given that the spec is very clear about the length of the arrays used and they are not unreasonably large to not use a definite length in most cases.

However, if we wanted to check up-front the length of indefinite-length arrays, we could use ReadEncodedValue() and then create another reader (new CborReader(encodedValue)) and use SkipValue() to count the elements. Then just throw away said reader and continue with the original one.